src/ray/object_manager/plasma/store.cc

```diff
@@ -231,8 +232,10 @@ uint8_t *PlasmaStore::AllocateMemory(size_t size, MEMFD_TYPE *fd, int64_t *map_s
     int64_t space_needed = eviction_policy_.RequireSpace(size, &objects_to_evict);
     EvictObjects(objects_to_evict);
     // More space is still needed. Try to spill objects to external storage to
-    // make room.
-    if (space_needed > 0) {
+    // make room. NOTE(ekl) if we can't achieve this after a number of retries,
+    // it's because memory fragmentation in dlmalloc prevents us from allocating
+    // even if our footprint tracker here still says we have free space.
```

Comment:
Why don't we use a different error for better error messages here? (e.g., FragmentationError?)

Reply:
I don't think we should be exposing these low-level details to the user.

Reply:
Hmm I agree with that. But would there be other good idea to see if this causes the OOM issue? I think this could help us debugging issues when it happens. Probably writing a log message?

Reply:
I don't think this causes any new OOM issues (previously, you'd get SIGBUS, now you'd get a proper error message or spilling).